[FLINK-4184] [metrics] Replace invalid characters in ScheduledDropwizardReporter

tillrohrmann opened the pull request:
The GraphiteReporter and GangliaReporter report metric names which can contain invalid
characters. These characters include quotes and dots. In order to properly report metrics
to these systems, the afore-mentioned characters have to be replaced in metric names.

The PR also removes quotes from the garbage collector metric name.

The PR sets the default value for TTL in the GangliaReporter to 1, because -1 causes the
reporter to fail.

R @zentol.

zentol commented on flink-core/src/main/java/org/apache/flink/metrics/reporter/AbstractReporter.java:
```diff
@@ -84,4 +84,24 @@ public void notifyOfRemovedMetric(Metric metric, String metricName, AbstractMetr
+			builder.append(replaceInvalidChars(componentName)).append(".");
+		}
+
+		builder.append(replaceInvalidChars(metricName));
```
the call to replaceInvalidChars is not required; metric names can only contain alphanumeric characters. I don't think any reporter will not support these.

zentol commented on flink-core/src/main/java/org/apache/flink/metrics/reporter/AbstractReporter.java:
```diff
@@ -84,4 +84,24 @@ public void notifyOfRemovedMetric(Metric metric, String metricName, AbstractMetr
+		StringBuilder builder = new StringBuilder();
+
+		for (String componentName : group.getScopeComponents()) {
+			builder.append(replaceInvalidChars(componentName)).append(".");
```
this is a bit inefficient. The output of this loop is identical for all metrics on that group, yet is computed for every single metric. Instead you could modify the getScopeString() method to accept a charFilter argument that is passed into ScopeFormat.concat().

zentol commented:
This PR will heavily conflict with #2219, in their current state we can't merge one without blocking the other.

tillrohrmann commented:
True, it conflicts with your proposed changes for the definable metric group delimiter. I will rebase and adapt this PR wrt #2219.

tillrohrmann commented:
I rebased on the latest master and introduced a `CharacterFilter` interface. The `CharacterFilter` allows to filter out invalid characters while generating the fully qualified metric name. 

In order to do this, the `AbstractMetricGroup#generateMetricName` takes a `CharacterFilter` as argument. The `AbstractReporter` and the `ScheduledDropwizardReporter` implement this interface to filter out reporter specific characters.

zentol commented on flink-core/src/main/java/org/apache/flink/metrics/groups/AbstractMetricGroup.java:
```diff
@@ -100,10 +101,29 @@ public AbstractMetricGroup(MetricRegistry registry, String[] scope) {
 		if (scopeString == null) {
-			scopeString = ScopeFormat.concat(registry.getDelimiter(), scopeComponents);
+			if (filter != null) {
+				scopeString = ScopeFormat.concat(registry.getDelimiter(), scopeComponents);
```
if no filter is given we will now never assign to scopeString, breaking all names. We should assume that no filtering is required.

tillrohrmann replied:
True, this is wrong. Fill fix it.

zentol commented:
do we have a test that verifies that reporters properly pass their filter when notified of new metrics?

zentol commented on flink-core/src/main/java/org/apache/flink/metrics/groups/AbstractMetricGroup.java:
```diff
@@ -100,10 +101,29 @@ public AbstractMetricGroup(MetricRegistry registry, String[] scope) {
+	 * {@code "host-7.taskmanager-2.window_word_count.my-mapper.metricName"}
+	 *
+	 * @param metricName metric name
+	 * @param filter character filter which is applied to the fully qualified metric name
```
this is misleading; it is not applied to the fully qualified name (as delimiter's are not filtered)

tillrohrmann replied:
Will adapt the comment

tillrohrmann commented:
Thanks for the review @zentol. I've addressed your comments and added test cases which verify that the corresponding `MetricReporters` filter out invalid characters in the fully qualified metric name.

zentol commented on flink-metrics/flink-metrics-dropwizard/src/main/java/org/apache/flink/dropwizard/ScheduledDropwizardReporter.java:
```diff
@@ -74,6 +75,15 @@ protected ScheduledDropwizardReporter() {
+	// ------------------------------------------------------------------------
+
+	// used for testing purposes
+	Map<Counter, String> getCounters() {
```
could we move this into the TestingScheduledDropwizardReporter?

tillrohrmann replied:
We can, if we mark the counters, gauges and histograms fields as protected. But then we would expose the implementation details to all sub-classes instead of having a getter which is package private. I think the latter option is a bit nicer, because it hides the implementation details.

zentol replied:
You could also access the protected registry and get the counters from there.

We may not even need the gauges/counters/histograms fields in the ScheduledDropwizardReporter.

zentol commented:
only one comment left, otherwise +1

tillrohrmann commented:
Thanks for the review @zentol. Will be merging this PR then.

zentol commented:
I would appreciate it if you would give me time to answer to your response before going ahead with a merge.
